src/best-practices.md

```diff
@@ -0,0 +1,173 @@
+# Best Practices
+
+## General Principles
+
+The main principle is simplification. Please keep it simple, and don't add new unnecessaries. In particular, see [Occam's razor](https://en.wikipedia.org/wiki/Occam%27s_razor#Software_development).
+
+### Convention Over Configuration
+
+If you can do something without configuring, e.g., using default settings, please do. Try to avoid introducing new "better" conventions. For more details, see [convention over configuration](https://en.wikipedia.org/wiki/Convention_over_configuration). Using fewer configurations reduces information noise.
+
+### Design Patterns And Anti-Patterns
+
+Don't worry much about [design patterns](https://en.wikipedia.org/wiki/Software_design_pattern), but avoid [anti-patterns](https://en.wikipedia.org/wiki/Anti-pattern). If you don't use some well-known design patterns, like the [factory method](https://en.wikipedia.org/wiki/Factory_method_pattern), it's okay. However, introducing an anti-pattern, such as a [mutable singleton](https://en.wikipedia.org/wiki/Singleton_pattern#Criticism), will damage our software's quality. 
+
+### No Direct Usage Of I/O
+
+Please avoid using direct I/O calls in libraries at all costs. You can bind direct I/O in the final application. Our libraries should never use direct I/O in the ideal world. I/O destroys deterministic behavior. Determinism is essential for testing and debugging because it allows us to reproduce behavior. If we can avoid I/O in our libraries, we can achieve 100% test coverage for them.
+
+A direct I/O call works like a virus. If a function `g` uses another function `f` that uses a direct I/O call, then `g` also uses a direct I/O call. Now, we have a huge challenge writing a unit test for the `g` function. Solution: If we can't change the `f` function, we may consider using a [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection).
+
+Try to avoid using multithreading and async programming in a library. Use simple synchronous functions that transform a state. A top-level application may decide how to bind the functions using different techniques, such as a [thread pool](https://en.wikipedia.org/wiki/Thread_pool). 
+
+## Documentation
+
+Use MarkDown to document a project. The GitHub MarkDown supports [Mermaid diagram](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams) and [LaTeX math equations](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions).
+
+## Git Workflow
+
+Use `main` branch as a default branch. The branch should be used for active development. Please don't create long-lived feature branches. You may create a long-living branch to create a fix for a critical bug in an old version of a package.
```

Comment:
I think this section could use some context as to why trunk-based development is the ideal vs gitflow. 
I can envision different projects that would benefit from each approach, so i feel adding some context here would be helpful. something like "if your code is consensus critical, gitflow may be a better and acceptable model to use vs trunk-based", where consensus-critical is just an example of where gitflow  may be more appropriate.

Reply:
Thank you for the feedback. It's as a good point. I think, the context is that we are trying to make it as simple as possible for developers and contributors. Use whatever GitHub/GitLab conventions are provided by default. 

I see one concern about using `main` for active development, that users may consume it directly from the branch. If we would like to ship our code as Rust crates, then it should not be a problem. Please, let me know if you know other problems related to active development in the `main` branch.

Reply:
I see your point, and it's a valid one.
I think this section is fine as it stands, and if we see a need in the future to adjust it - we can do so then.